### PR TITLE
ClearMLCallback enhancements: support multiple runs and handle logging better

### PR DESCRIPTION
**Pull request by eugen-ajechiloae-clearml**
Currently, training multiple models in the same script might cause some model checkpoints logged to ClearML to be lost, as well as scalars, when using ClearMLCallback. This PR fixes these issues. Scalar visualization in ClearML has also been enhanced. 

What the PR does:
1. We count the number of times `ClearMLCallback.setup` is called via class variables. When a second training run is created, we do one of the following: we create a new task that will be used to log all the models, metrics etc. OR we keep the same task and we suffix the metrics/checkpoint etc. with the setup number.
We keep the same task if the task was created externally or if we are running remotely. We create new tasks if `ClearMLCallback` is the one that created the first task (in this case we also close the task so we can create another one).
2. We now delete model checkpoints if `save_total_limit` is set and the limit has been exceeded.
3. We switched the title/series of the logged scalars for better visualization.
4. We now, by default, don't fetch the configurations/hparams from the backend when running remotely, as these can contain temp files or other variables that are related to the local environment. The user can still override these tho, by setting, in the UI/via scripts, `_ignore_hparams_ui_overrides_` or `_ignore_model_config_ui_overrides_` to False.

### CODE REVIEW

**Comment by eugen-ajechiloae-clearml** [comment]
Hi @LysandreJik! Can you or someone from your team please review this PR?

**Comment by amyeroberts** [comment]
Hi @eugen-ajechiloae-clearml, apologies for the delay. Reviewing now!

File: src/transformers/integrations/integration_utils.py [hunk 1]
```diff
@@ -1543,7 +1653,40 @@ def on_save(self, args, state, control, **kwargs):
             ckpt_dir = f"checkpoint-{state.global_step}"
             artifact_path = os.path.join(args.output_dir, ckpt_dir)
             logger.info(f"Logging checkpoint artifacts in {ckpt_dir}. This may take time.")
```
Comment (amyeroberts) [inline src/transformers/integrations/integration_utils.py:1655]: This message should be update to reflect saving under `name`

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding these improvements! 

Just a few small questions and notes

File: src/transformers/integrations/integration_utils.py [hunk 2]
```diff
@@ -1543,7 +1653,40 @@ def on_save(self, args, state, control, **kwargs):
...
+                        )
+                    )
+                    break
+                self._checkpoints_saved = self._checkpoints_saved[1:]
+
+    def _copy_training_args_as_hparams(self, training_args, prefix):
+        as_dict = {field.name: getattr(training_args, field.name) for field in fields(training_args) if field.init}
+        token_keys = [k for k in as_dict.keys() if k.endswith("_token")]
+        for token_key in token_keys:
+            as_dict.pop(token_key, None)
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/transformers/integrations/integration_utils.py]: done

File: src/transformers/integrations/integration_utils.py [hunk 3]
```diff
@@ -1447,54 +1465,135 @@ def __init__(self):
...
                     logger.info("External ClearML Task has been connected.")
                 else:
                     self._clearml_task = self._clearml.Task.init(
                         project_name=os.getenv("CLEARML_PROJECT", "HuggingFace Transformers"),
                         task_name=os.getenv("CLEARML_TASK", "Trainer"),
-                        auto_connect_frameworks={"tensorboard": False, "pytorch": False},
+                        auto_connect_frameworks={
+                            "tensorboard": False,
+                            "pytorch": False,
+                        },
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/transformers/integrations/integration_utils.py]: done

File: src/transformers/integrations/integration_utils.py [hunk 4]
```diff
@@ -1447,54 +1465,135 @@ def __init__(self):
...
+                    ClearMLCallback._task_created_in_callback = True
                     logger.info("ClearML Task has been initialized.")
+                self._initialized = True
 
-            self._clearml_task.connect(args, "Args")
-            if hasattr(model, "config") and model.config is not None:
-                self._clearml_task.connect(model.config, "Model Configuration")
+            ignore_hparams_config_section = (
+                ClearMLCallback._hparams_section
+                + ClearMLCallback.log_suffix
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/transformers/integrations/integration_utils.py]: done

File: src/transformers/integrations/integration_utils.py (hunk 1 above)
Comment (eugen-ajechiloae-clearml) [reply src/transformers/integrations/integration_utils.py]: done

File: src/transformers/integrations/integration_utils.py [hunk 5]
```diff
@@ -1447,54 +1465,135 @@ def __init__(self):
...
+                        {"TRUE"}
+                    )
+                    ClearMLCallback._task_created_in_callback = True
                     logger.info("ClearML Task has been initialized.")
+                self._initialized = True
 
-            self._clearml_task.connect(args, "Args")
-            if hasattr(model, "config") and model.config is not None:
-                self._clearml_task.connect(model.config, "Model Configuration")
+            ignore_hparams_config_section = (
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/transformers/integrations/integration_utils.py]: It does actually represent a boolean flag, but the flag is actually set in a config section in the UI. The description of this flag can be found here for example: https://github.com/eugen-ajechiloae-clearml/transformers/blob/6dca78e722e461aa28a9afe229efc96e2931c240/src/transformers/integrations/integration_utils.py#L1523

File: src/transformers/integrations/integration_utils.py [hunk 6]
```diff
@@ -1447,54 +1465,135 @@ def __init__(self):
...
+                )
+            elif not self._clearml_task.get_parameter(ignore_hparams_config_section, default=True, cast=True):
+                self._clearml_task.connect(args, ClearMLCallback._hparams_section + ClearMLCallback.log_suffix)
+            else:
+                self._copy_training_args_as_hparams(
+                    args, ClearMLCallback._hparams_section + ClearMLCallback.log_suffix
+                )
+
+            if getattr(model, "config", None) is not None:
+                ignore_model_config_section = (
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/transformers/integrations/integration_utils.py:1536]: Same as: https://github.com/huggingface/transformers/pull/28559#discussion_r1472811196

**Review by amyeroberts — APPROVED** [review]
Thanks for improving this integration!

**Comment by amyeroberts** [comment]
@eugen-ajechiloae-clearml We had some recent issues with new package releases and compatibility on our circle CI pipeline. Fixes have been merged into `main` rebasing and pushing the changes should resolve the tests that are currently failing 🤗

**Comment by eugen-ajechiloae-clearml** [comment]
@amyeroberts I believe I synced the branch appropriately. Is there something I'm missing?

**Comment by amyeroberts** [comment]
@eugen-ajechiloae-clearml I'm not sure what's happening to be honest. From the commit history, it all looks good. And I'm pretty sure these failures are independent of the changes in this PR. I'm going to see if this is popping up elsewhere.

**Comment by amyeroberts** [comment]
OK, seems like there's some outstanding issues with package installs. There's a PR open to fix: #28834. Once that's merged in, a final rebase should do it :)

**Comment by dimka11** [comment]
This new implementation ClearML callback's output all metrics from log as multiple series at one plot, that's looks not readable. 
image
For example: runtime and samples_per_seconds has different scale than other metrics. 
Any ideas how can  we set / filter metrics without code changes?
As temporarily solution I add my own callback for send some metrics to ClearML at other plots.

**Pull request merged**